raiden-ts/src/channels/epics.ts

```diff
@@ -289,145 +282,145 @@ export const channelMonitoredEpic = (
-                      action.meta,
-                    ),
-                  ),
-                ),
-                getEventsStream<ChannelSettledEvent>(
-                  tokenNetworkContract,
-                  [settledFilter],
-                  action.payload.fromBlock ? of(action.payload.fromBlock) : undefined,
-                  action.payload.fromBlock ? blockNumber$ : undefined,
-                ).pipe(
-                  map(([id, , , , , event]) =>
-                    channelSettled(
-                      {
-                        id: id.toNumber(),
-                        settleBlock: event.blockNumber!,
-                        txHash: event.transactionHash! as Hash,
-                      },
-                      action.meta,
-                    ),
-                  ),
-                ),
-              ).pipe(
-                // takeWhile tends to broad input to generic Action. We need to narrow it by hand
-                takeWhile<
-                  ActionType<
-                    | typeof channelDeposited
-                    | typeof channelWithdrawn
-                    | typeof channelClosed
-                    | typeof channelSettled
-                  >
-                >(negate(isActionOf(channelSettled)), true),
-              );
+          /**
+           * Guards that an event data tuple matches the type of a given filter
+           *
+           * Type must be explicitly passed as generic type parameter, and a corresponding filter
+           * as first parameter
+           *
+           * @param filter - Filter of an event of type T
+           * @param data - event data tuple, where last element is the Event object
+           * @returns Truty if event data matches filter
+           */
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          function isEvent<T extends any[]>(filter: Filter, data: any[]): data is T {
+            const event = data[data.length - 1] as Event;
+            if (!event || !event.topics || !filter.topics) return false;
+            const topic0 = filter.topics[0];
+            return Array.isArray(topic0)
+              ? topic0.includes(event.topics[0])
+              : topic0 === event.topics[0];
+          }
+
+          return getEventsStream<
+            | ChannelNewDepositEvent
+            | ChannelWithdrawEvent
+            | ChannelClosedEvent
+            | ChannelSettledEvent
+          >(
+            tokenNetworkContract,
+            [mergedFilter],
+            // if channelMonitored triggered by ChannelOpenedAction,
+            // fetch Channel's pastEvents since channelOpened blockNumber as fromBlock$
+            action.payload.fromBlock ? of(action.payload.fromBlock) : undefined,
+          ).pipe(
+            mergeMap(function*(data) {
+              if (isEvent<ChannelNewDepositEvent>(depositFilter, data)) {
+                yield channelDeposited(
+                  {
+                    id: data[0].toNumber(),
```

Comment:
Maybe worth destructuring for better readability:
```
const [id, participant, transaction, block, ...] = data;
```

Reply:
Would need to do separately each block for all parameters, do you think it's worth it if they're used only once?

Reply:
Only for readability. Feel free to ignore if you think it's not worth it